cleanlab/datalab/internal/issue_manager/duplicate.py

```diff
@@ -76,32 +75,11 @@ def find_issues(
         metric_changes = self.metric and self.metric != old_knn_metric
 
         if knn_graph is None or metric_changes:
-            if features is None:
-                raise ValueError(
-                    "If a knn_graph is not provided, features must be provided to fit a new knn."
-                )
-            if self.metric is None:
-                self.metric = (
-                    "cosine"
-                    if features.shape[1] > 3
-                    else "euclidean" if features.shape[0] > 100 else euclidean
-                )
-            knn = NearestNeighbors(n_neighbors=self.k, metric=self.metric)
-
-            if self.metric and self.metric != knn.metric:
-                warnings.warn(
-                    f"Metric {self.metric} does not match metric {knn.metric} used to fit knn. "
-                    "Most likely an existing NearestNeighbors object was passed in, but a different "
-                    "metric was specified."
-                )
-            self.metric = knn.metric
-
-            try:
-                check_is_fitted(knn)
-            except NotFittedError:
-                knn.fit(features)
-
-            knn_graph = knn.kneighbors_graph(mode="distance")
+            _features = cast(np.ndarray, features)
+            self.metric = self.metric or decide_default_metric(_features)
```

Comment:
it seems like this:
`self.metric = self.metric or decide_default_metric(_features)`
is always being used before `construct_knn_graph_from_features()` 

in every IssueManager?  Why not move this line of code inside of the `construct_knn_graph_from_features()` function?

Reply:
same with the `cast(np.ndarray, features)` call, which appears to be happening in many of the IssueManagers, and could perhaps be inside of the `construct_knn_graph_from_features()` function. I'm not sure if the cast is supposed to have longer-term effects on `features` though that persist outside of this function...

Reply:
The `self.metric = ...` turned out to be necessary when `construct_knn_graph_from_features()` only returned a `csr_matrix` (a sparse matrix), while most of the issue managers wanted to track the metric that was set for the `knn` object.

The line:

```python
self.metric = self.metric or decide_default_metric(_features)
```
was already inside of the `construct_knn_graph_from_features()` (specifically the `features_to_knn()` function that constructed the `knn` object).

I've now updated the function to return both the spare matrix as well as the `knn` metric, which we can assume to have a `metric` attribute.

---

Regarding the `cast()`, it has no runtime impact, just a way to ensure the type-checker that at this point in the code, that features is considered a `np.ndarray`. This is now redundant after I removed `self.metric` lines, because `construct_knn_graph_from_features()` properly raises a runtime error when `features = None`.